File: apps/mobile/modules/location-engine/ios/LocationEngineModule.swift
import ExpoModulesCore

/**
 * Expo Modules bridge for the iOS location engine.
 *
 * This file is only a bridge: it validates and marshals, then delegates to
 * `LocationEngine`. Keeping policy out of here means the engine can be unit
 * tested in Swift without the Expo runtime.
 *
 * Every method that could start collecting location is asynchronous and can
 * fail. Nothing here silently succeeds — a caller that cannot start tracking
 * must find out, so the UI can tell the user that sharing is not running.
 */
public final class LocationEngineModule: Module {
    private let engine = LocationEngine.shared

    public func definition() -> ModuleDefinition {
        Name("LocationEngineModule")

        Events(
            "onEngineStateChange",
            "onPermissionChange",
            "onGeofenceTransition",
            "onQueueChange"
        )

        OnCreate {
            engine.eventSink = { [weak self] name, payload in
                self?.sendEvent(name, payload)
            }
        }

        OnDestroy {
            engine.eventSink = nil
        }

        AsyncFunction("configure") { (config: [String: Any]) in
            try engine.configure(EngineConfiguration(dictionary: config))
        }

        AsyncFunction("startPassiveTracking") {
            try engine.startPassiveTracking()
        }

        AsyncFunction("startLiveSession") { (session: [String: Any]) in
            guard
                let sessionId = session["sessionId"] as? String,
                let duration = session["durationSeconds"] as? Double
            else {
                throw InvalidArgumentException("startLiveSession requires sessionId and durationSeconds")
            }
            let interval = session["updateIntervalSeconds"] as? Double
            try engine.startLiveSession(
                sessionId: sessionId,
                durationSeconds: duration,
                updateIntervalSeconds: interval
            )
        }

        AsyncFunction("stopLiveSession") {
            engine.stopLiveSession(reason: .stoppedByUser)
        }

        AsyncFunction("pauseSharing") {
            engine.pauseSharing()
        }

        AsyncFunction("resumeSharing") {
            try engine.resumeSharing()
        }

        AsyncFunction("getPermissionState") { () -> [String: Any?] in
            engine.permissionState()
        }

        AsyncFunction("getDeviceHealth") { () -> [String: Any?] in
            engine.deviceHealth()
        }

        AsyncFunction("flushPendingEvents") { () -> [String: Any?] in
            engine.flushPendingEvents()
        }

        AsyncFunction("registerGeofences") { (places: [[String: Any]]) in
            try engine.registerGeofences(places.compactMap(GeofenceRegion.init(dictionary:)))
        }

        AsyncFunction("unregisterGeofences") { (placeIds: [String]) in
            engine.unregisterGeofences(placeIds: placeIds)
        }
    }
}

/// Thrown when JavaScript hands the bridge something the engine cannot use.
///
/// `GenericException` is the Expo Modules pattern for carrying a payload into
/// the message. Subclassing `Exception` directly and adding a stored `reason`
/// collides with the base class's own computed property of that name.
final class InvalidArgumentException: GenericException<String> {
    override var reason: String {
        param
    }
}

File: apps/mobile/modules/location-engine/android/src/main/java/com/familylocation/locationengine/LocationEngineModule.kt
package com.familylocation.locationengine

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Expo Modules bridge for the Android location engine.
 *
 * A bridge only: it marshals arguments and delegates to `LocationEngine`, so
 * the engine stays testable without the Expo runtime.
 *
 * Every method that could start collecting location can fail, and failures
 * propagate to JavaScript. Nothing silently succeeds — a caller that cannot
 * start tracking must find out, so the UI can tell the user sharing is off.
 */
class LocationEngineModule : Module() {

    private val engine: LocationEngine
        get() = LocationEngine.getOrCreate(
            appContext.reactContext ?: throw MissingContextException(),
        )

    override fun definition() = ModuleDefinition {
        Name("LocationEngineModule")

        Events(
            "onEngineStateChange",
            "onPermissionChange",
            "onGeofenceTransition",
            "onQueueChange",
        )

        OnCreate {
            engine.eventSink = { name, payload -> sendEvent(name, payload) }
        }

        OnDestroy {
            LocationEngine.instance?.eventSink = null
        }

        AsyncFunction("configure") { config: Map<String, Any?> ->
            engine.configure(EngineConfiguration.fromMap(config))
        }

        AsyncFunction("startPassiveTracking") {
            runOrThrow("startPassiveTracking") { engine.startPassiveTracking() }
        }

        AsyncFunction("startLiveSession") { session: Map<String, Any?> ->
            val sessionId = session["sessionId"] as? String
                ?: throw InvalidArgumentException("startLiveSession requires sessionId")
            val duration = (session["durationSeconds"] as? Number)?.toLong()
                ?: throw InvalidArgumentException("startLiveSession requires durationSeconds")
            runOrThrow("startLiveSession") { engine.startLiveSession(sessionId, duration) }
        }

        AsyncFunction("stopLiveSession") { engine.stopLiveSession() }

        AsyncFunction("pauseSharing") { engine.pauseSharing() }

        AsyncFunction("resumeSharing") {
            runOrThrow("resumeSharing") { engine.resumeSharing() }
        }

        AsyncFunction("getPermissionState") { engine.permissionState() }

        AsyncFunction("getDeviceHealth") { engine.health() }

        AsyncFunction("flushPendingEvents") { engine.pendingBatch() }

        AsyncFunction("confirmUploaded") { eventIds: List<String> ->
            engine.confirmUploaded(eventIds)
        }

        AsyncFunction("registerGeofences") { places: List<Map<String, Any?>> ->
            engine.registerGeofences(places)
        }

        AsyncFunction("unregisterGeofences") { placeIds: List<String> ->
            engine.unregisterGeofences(placeIds)
        }
    }

    /**
     * Converts the engine's `check(...)` preconditions into coded exceptions so
     * JavaScript gets a stable, user-safe reason rather than a stack trace.
     */
    private inline fun runOrThrow(operation: String, block: () -> Unit) {
        try {
            block()
        } catch (error: IllegalStateException) {
            throw EngineStateException(operation, error.message ?: "precondition failed")
        }
    }
}

internal class MissingContextException :
    CodedException("The Android context is not available; the module is not attached.")

internal class InvalidArgumentException(reason: String) : CodedException(reason)

internal class EngineStateException(operation: String, reason: String) :
    CodedException("$operation could not run: $reason")
